File: APPLICATION/services/producto_service.py
"""
producto_service.py — Servicio CRUD para productos.
Toda persistencia se delega al FileManager (struct + seek()).
"""
from typing import List, Optional

from models.producto import Producto
from services.file_manager import FileManager
from utils.config import PRODUCTOS_DAT, PRODUCTOS_IDX, PRODUCTOS_HOLES
from utils.logger import get_logger

logger = get_logger("producto_service")


class ProductoService:
    """Operaciones CRUD sobre productos usando archivos binarios de acceso directo."""

    def __init__(self) -> None:
        self._fm = FileManager(
            dat_path=PRODUCTOS_DAT,
            idx_path=PRODUCTOS_IDX,
            holes_path=PRODUCTOS_HOLES,
            model_class=Producto,
        )

    # ── Crear ────────────────────────────────────────────────────────────────

    def crear(self, nombre: str, precio: float, stock: int, categoria: str) -> Producto:
        """Registra un nuevo producto en el inventario."""
        if not nombre.strip():
            raise ValueError("El nombre del producto no puede estar vacío.")
        if precio < 0:
            raise ValueError("El precio no puede ser negativo.")
        if stock < 0:
            raise ValueError("El stock no puede ser negativo.")

        producto = Producto(nombre=nombre.strip(), precio=precio,
                            stock=stock, categoria=categoria.strip(), estado=1)
        self._fm.insert(producto)
        logger.info(f"Producto creado: {producto}")
        return producto

    # ── Leer ─────────────────────────────────────────────────────────────────

    def obtener(self, producto_id: int) -> Optional[Producto]:
        """Acceso directo O(1) por ID."""
        return self._fm.find_by_id(producto_id)

    def listar(self) -> List[Producto]:
        """Todos los productos activos."""
        return self._fm.find_all()

    def buscar(self, termino: str) -> List[Producto]:
        """Búsqueda por nombre o categoría (filtra sobre resultados del índice)."""
        t = termino.lower().strip()
        return [
            p for p in self._fm.find_all()
            if t in p.nombre.lower() or t in p.categoria.lower()
        ]

    def obtener_por_ids(self, ids: List[int]) -> List[Producto]:
        """Carga múltiples productos por lista de IDs."""
        result = []
        for pid in ids:
            p = self._fm.find_by_id(pid)
            if p:
                result.append(p)
        return result

    # ── Actualizar ───────────────────────────────────────────────────────────

    def actualizar(
        self,
        producto_id: int,
        nombre:    Optional[str]   = None,
        precio:    Optional[float] = None,
        stock:     Optional[int]   = None,
        categoria: Optional[str]   = None,
    ) -> Optional[Producto]:
        """Modifica campos de un producto existente (escritura in-place)."""
        prod = self._fm.find_by_id(producto_id)
        if prod is None:
            logger.warning(f"Producto id={producto_id} no encontrado para actualizar")
            return None

        if nombre    is not None: prod.nombre    = nombre.strip()
        if precio    is not None: prod.precio    = precio
        if stock     is not None: prod.stock     = stock
        if categoria is not None: prod.categoria = categoria.strip()

        self._fm.update(prod)
        logger.info(f"Producto actualizado: {prod}")
        return prod

    def descontar_stock(self, producto_id: int, cantidad: int) -> bool:
        """
        Descuenta 'cantidad' unidades del stock al procesar una venta.
        Retorna False si el stock es insuficiente.
        """
        prod = self._fm.find_by_id(producto_id)
        if prod is None:
            logger.error(f"Producto id={producto_id} no encontrado para descontar stock")
            return False
        if prod.stock < cantidad:
            logger.warning(
                f"Stock insuficiente: id={producto_id} "
                f"disponible={prod.stock} solicitado={cantidad}"
            )
            return False
        prod.stock -= cantidad
        return self._fm.update(prod)

    def restaurar_stock(self, producto_id: int, cantidad: int) -> bool:
        """Restaura stock al anular una venta."""
        prod = self._fm.find_by_id(producto_id)
        if prod is None:
            return False
        prod.stock += cantidad
        return self._fm.update(prod)

    # ── Eliminar ─────────────────────────────────────────────────────────────

    def eliminar(self, producto_id: int) -> bool:
        """Eliminación lógica (estado=0). El slot queda disponible para reutilizar."""
        ok = self._fm.delete(producto_id)
        if ok:
            logger.info(f"Producto id={producto_id} eliminado lógicamente")
        return ok

    # ── Utilidades ───────────────────────────────────────────────────────────

    def contar(self) -> int:
        return self._fm.count_active()

    def compactar(self) -> int:
        """Compactación física del archivo .dat."""
        return self._fm.compact()

    def exportar_csv(self, export_path) -> int:
        """Exporta todos los productos activos a CSV para sincronización."""
        productos = self.listar()
        with open(export_path, "w", encoding="utf-8") as f:
            f.write("id,nombre,precio,stock,categoria,estado\n")
            for p in productos:
                f.write(p.to_csv_row() + "\n")
        logger.info(f"Exportados {len(productos)} productos a {export_path}")
        return len(productos)

File: APPLICATION-JAVA/src/main/java/com/minimarket/services/ProductoService.java
package com.minimarket.services;

import com.minimarket.models.Producto;
import com.minimarket.utils.Config;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.logging.Logger;
import java.util.stream.Collectors;

/**
 * Servicio CRUD para productos.
 * Toda la persistencia se delega al FileManager (RandomAccessFile + seek()).
 */
public class ProductoService {

    private static final Logger LOG = Logger.getLogger(ProductoService.class.getName());

    private final FileManager<Producto> fm;

    public ProductoService() {
        fm = new FileManager<>(
            Config.PRODUCTOS_DAT,
            Config.PRODUCTOS_IDX,
            Config.PRODUCTOS_HOLES,
            Producto.SIZE,
            Producto::fromBytes,
            "Producto"
        );
    }

    // ── Crear ────────────────────────────────────────────────────────────────

    public Producto crear(String nombre, double precio, int stock, String categoria) {
        if (nombre == null || nombre.isBlank())
            throw new IllegalArgumentException("El nombre del producto no puede estar vacío.");
        if (precio < 0)
            throw new IllegalArgumentException("El precio no puede ser negativo.");
        if (stock < 0)
            throw new IllegalArgumentException("El stock no puede ser negativo.");

        Producto producto = new Producto(0, nombre.strip(), precio, stock, categoria.strip(), 1);
        fm.insert(producto);
        LOG.info("Producto creado: " + producto);
        return producto;
    }

    // ── Leer ─────────────────────────────────────────────────────────────────

    /** Acceso directo O(1) por ID. */
    public Producto obtener(int productoId) {
        return fm.findById(productoId);
    }

    /** Todos los productos activos. */
    public List<Producto> listar() {
        return fm.findAll();
    }

    /** Busqueda por nombre o categoria (filtro sobre resultados del indice). */
    public List<Producto> buscar(String termino) {
        String t = termino.toLowerCase().strip();
        return fm.findAll().stream()
            .filter(p -> p.getNombre().toLowerCase().contains(t)
                      || p.getCategoria().toLowerCase().contains(t))
            .collect(Collectors.toList());
    }

    // ── Actualizar ───────────────────────────────────────────────────────────

    public Producto actualizar(int productoId, String nombre, Double precio,
                               Integer stock, String categoria) {
        Producto prod = fm.findById(productoId);
        if (prod == null) {
            LOG.warning("Producto id=" + productoId + " no encontrado para actualizar");
            return null;
        }
        if (nombre    != null) prod.setNombre(nombre.strip());
        if (precio    != null) prod.setPrecio(precio);
        if (stock     != null) prod.setStock(stock);
        if (categoria != null) prod.setCategoria(categoria.strip());

        fm.update(prod);
        LOG.info("Producto actualizado: " + prod);
        return prod;
    }

    /**
     * Descuenta 'cantidad' unidades del stock al procesar una venta.
     * @return false si el stock es insuficiente.
     */
    public boolean descontarStock(int productoId, int cantidad) {
        Producto prod = fm.findById(productoId);
        if (prod == null) {
            LOG.severe("Producto id=" + productoId + " no encontrado para descontar stock");
            return false;
        }
        if (prod.getStock() < cantidad) {
            LOG.warning("Stock insuficiente: id=" + productoId
                + " disponible=" + prod.getStock() + " solicitado=" + cantidad);
            return false;
        }
        prod.setStock(prod.getStock() - cantidad);
        return fm.update(prod);
    }

    /** Restaura stock al anular una venta. */
    public boolean restaurarStock(int productoId, int cantidad) {
        Producto prod = fm.findById(productoId);
        if (prod == null) return false;
        prod.setStock(prod.getStock() + cantidad);
        return fm.update(prod);
    }

    // ── Eliminar ─────────────────────────────────────────────────────────────

    /** Eliminacion logica (estado=0). El slot queda disponible para reutilizar. */
    public boolean eliminar(int productoId) {
        boolean ok = fm.delete(productoId);
        if (ok) LOG.info("Producto id=" + productoId + " eliminado logicamente");
        return ok;
    }

    // ── Utilidades ───────────────────────────────────────────────────────────

    public int contar() { return fm.countActive(); }

    public int compactar() { return fm.compact(); }

    /** Exporta todos los productos activos a CSV para sincronizacion. */
    public int exportarCsv(Path exportPath) throws IOException {
        List<Producto> productos = listar();
        try (PrintWriter pw = new PrintWriter(Files.newBufferedWriter(exportPath, StandardCharsets.UTF_8))) {
            pw.println("id,nombre,precio,stock,categoria,estado");
            for (Producto p : productos) {
                pw.println(p.toCsvRow());
            }
        }
        LOG.info("Exportados " + productos.size() + " productos a " + exportPath);
        return productos.size();
    }
}
